src/events/InputEventGenerator.cc

```diff
@@ -312,6 +312,20 @@ void InputEventGenerator::handle(const SDL_Event& evt)
 		break;
 	case SDL_MOUSEMOTION:
 		event = MouseMotionEvent(evt);
+		{
+			auto *window = SDL_GL_GetCurrentWindow();
+			if (SDL_GetWindowGrab(window)) {
+				int x, y, xn, yn, w, h;
+				SDL_GetWindowSize(window, &w, &h);
+				SDL_GetMouseState(&x, &y);
+				xn = x; yn = y;
+				if (x <= 10)     xn = 10;
+				if (x >= w - 2)  xn = w/2;
+				if (y <= 10)     yn = 10;
+				if (y >= h - 2)  yn = h/2;
+				if (xn != x || yn != y) SDL_WarpMouseInWindow(window, xn, yn);
```

Comment:
Yes, this code has grown out of experiment.  Ideally, I'd like to always test edges (x,y <= 1 && x,y >= w,h - 2) and warp to the middle of the windows.  However when doing that with y, there is often interference with the menubar appearing: sometimes it starts appearing and yet the mouse warps away, annoyingly needing another attempt.  For x that would not be necessary, but there I wanted to create at least a bit of symmetry.  The truth is however that I had first done the other way around (x,y <= 10 && x,y <= w,h - 11) but for some reason now horizontally the mouse could hang on the last pixel (and even become visible there).  I tried to increase the margin, but to no avail, and being tired of trying to finger this yet another strange anomaly, I decided to go for the middle of the window (and have symmetry).  So, in short: y needs small edge, x needs large edge.

Reply:
I'm sorry but I'm still confused.
* What I meant is that left vs right (or top vs bottom) symmetry is broken in the code. But you say there's a good reason for that? It would be very useful to document such things in the code. Because such knowledge is currently impossible to retrieve.
* And not only are the values different (`10` for left margin, `2` for right margin), but also the behavior itself: left is clamped to `10`, right is warped to the middle. That's also very unexpected behavior and (if intentional) should be documented.
* The current code is symmetrical for X vs Y. But from what I understood from your explanation the real behavior for X and Y is NOT symmetrical. If so, then maybe we shouldn't artificially make it symmetrical in the code?

May I ask to treat the 4 cases (left, right, top, bottom) separately. Start with a common simple implementation for all 4, but then individually introduce exceptions. And for each exception clearly document why it's needed.

Or of course if there is a common implementation that works for all 4 (e.g. unconditionally always warp to the middle?) that would be even better.

Reply:
Ok, let me try to explain better ...

Naively, the first implementation would be to detect that the mouse is at the penultimate pixel at any edge and warp it just one pixel backwards.  This works, but makes for a very slow compensation when continuing the movement in the forward direction.

So, a bigger margin is needed.  It turns out 10 pixels is a good heuristic.  Now the movement to compensate for virtual barriers can be processed fast enough.  Now to translate this into numbers, be careful with width and height.  The first pixel is of course x=0 or y=0, but the last pixel is x=w-1 or y=h-1.  So this already introduces a bit of asymmetry.  So for the margin this becomes x=1, y=1 and x=w-2 and y=h-2.  For the margin this becomes x=10, y=10 and x=w-11, y=h-11.

This would be good.  Warping it further towards the middle of the window will probably improve the situation, yet only marginally, and would introduce the issue with the interrupted attempt to raise the menubar (as described previously).

No enter this strange anomaly where at right edge, moving towards the right, the Windows mouse would suddenly become visible and block at the last pixel (because of the grabbing).  In that state, moving the mouse up or down does no longer cause mouse events.  The mouse can go back to the left, becomes invisible again and generate mouse events.  But this is a very weird mess-up again.  I thought I had miscalculated the edge (with the -1 or -2) but increasing the edge to w-11 or even w-22 or so did not solve it.  Brute forcing it to w/2 did solve it.  As I said I did not feel like fingering this one out again, or try to find the right interpolation value.  Asymmetry would be the result anyway, so w/2 is as good as any then, and has some esthetic value at least.

It is possible that the same anomaly does not happen at the bottom edge, because of the activation of ImGui which happens there (because of the leds and the timeline).  So the only question left is: do we also go for y=h/2 or do we keep it at y=h-11?  Or else we need to understand why this anomaly ... and then keep everything at 10 pixels margin ...

Reply:
PS: I wonder now if ImGui does not occupy the right edge, and hence this behavior?  I am not sure because when at the bottom ImGui is activated (e.g. over the leds or when the timeline raises) also the menubar raises, and I don't remember that this happened in the case of the right edge anomaly ...  I'd need to check again ...

Reply:
For your experiments you could try disabling the leds and reverse bar. Of course the final solution should also work when they are enabled.

Thanks for the extra explanation. I think it's valuable to document such weird quirks/exceptions in the code. One possibility is to add a comment referring to this discussion (at least I think you can get a stable URL for these individual replies?).